### PR TITLE
Custom swap accounting with NoOps

### PR DESCRIPTION
**Pull request by emmaguo13**
Inspired by NoOp design from https://github.com/Uniswap/v4-core/pull/291. @thogard785's commit is merged with this PR

### CODE REVIEW

File: contracts/PoolManager.sol [hunk 1]
```diff
@@ -225,8 +257,23 @@ contract PoolManager is IPoolManager, Owned, NoDelegateCall, ERC1155, IERC1155Re
...
+                    _accountNoOp(key, hookDelta);
+                    if (key.hooks.shouldCallAfterModifyPosition()) {
+                        if (
+                            key.hooks.afterModifyPosition(msg.sender, key, params, delta)
+                                != IHooks.afterModifyPosition.selector
+                        ) {
+                            revert Hooks.InvalidHookResponse();
+                        }
+                    }
+                    return hookDelta;
```
Comment (NoahZinsmeister) [inline contracts/PoolManager.sol:273]: i guess because it makes this return a little redundant...

File: contracts/types/BalanceDelta.sol [hunk 2]
```diff
@@ -1,6 +1,8 @@
 // SPDX-License-Identifier: GPL-2.0-or-later
 pragma solidity ^0.8.20;
 
+import "forge-std/console2.sol";
```
Comment (NoahZinsmeister) [inline contracts/types/BalanceDelta.sol:4]: unused

File: contracts/types/BalanceDelta.sol [hunk 3]
```diff
@@ -14,6 +16,23 @@ function toBalanceDelta(int128 _amount0, int128 _amount1) pure returns (BalanceD
...
+        amt0 := and(0x000000000000000000000000000000000000ffffffffffffffffffffffffffff, input)
+        amt1 := and(0x000000000000000000000000000000000000ffffffffffffffffffffffffffff, shr(112, input))
+    }
+    int128 _amount0 = int128(amt0);
+    int128 _amount1 = int128(amt1);
+
+    assembly {
+        balanceDelta :=
+            or(shl(128, _amount0), and(0x00000000000000000000000000000000ffffffffffffffffffffffffffffffff, _amount1))
+    }
```
Comment (NoahZinsmeister) [inline contracts/types/BalanceDelta.sol:33]: this should just be `toBalanceDelta`

**Review by NoahZinsmeister — CHANGES_REQUESTED** [review]
i wonder if it's worth allowing pools to use a returned hooksDelta even in non-NO_OP cases...one reason is i think it's awkward if initialize can NO_OP (for example, sqrtPriceX96 remains 0 in this case, which i think means the pool is borked unless it exclusively uses no-ops going forward, which is fine for some pools but maybe not for others, and feels like a bit of a footgun). so, i think we should consider disallowing no-op on initialize, but the delta logic could still be useful for hooks to levy an initial fee or incentivize the creator or something?

i also don't really think we should allow calling `after*` hooks when NO_OP is used

finally i think a version of `_accountNoOp` that applies fees would be valuable to see. for the sake of clarity it might make sense to write that as a stacked PR on top of this one

**Comment by emmaguo13** [comment]
NoOping on initialize will prevent the pool with that hook from being included in our pools mapping, so I think it should be removed.

File: contracts/test/GracefulReturnTestHooks.sol [hunk 4]
```diff
@@ -0,0 +1,238 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.19;
+
+import {Hooks} from "../libraries/Hooks.sol";
+import {IHooks} from "../interfaces/IHooks.sol";
+import {IPoolManager} from "../interfaces/IPoolManager.sol";
+import {BalanceDelta, toBalanceDelta} from "../types/BalanceDelta.sol";
```
Comment (0xb10ckdev) [inline contracts/test/GracefulReturnTestHooks.sol:7]: nit: unused `toBalanceDelta` import

File: contracts/test/NoOpTestHooks.sol [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+import {Hooks} from "../libraries/Hooks.sol";
+import {IHooks} from "../interfaces/IHooks.sol";
+import {IPoolManager} from "../interfaces/IPoolManager.sol";
+import {BalanceDelta, toBalanceDelta} from "../types/BalanceDelta.sol";
+import {IHookFeeManager} from "../interfaces/IHookFeeManager.sol";
+import {PoolId, PoolIdLibrary} from "../libraries/PoolId.sol";
+import {IERC1155Receiver} from "@openzeppelin/contracts/token/ERC1155/IERC1155Receiver.sol";
+import {CurrencyLibrary, Currency} from "../libraries/CurrencyLibrary.sol";
+import {IERC20Minimal} from "../interfaces/external/IERC20Minimal.sol";
+import {ILockCallback} from "../interfaces/callback/ILockCallback.sol";
```
Comment (0xb10ckdev) [inline contracts/test/NoOpTestHooks.sol:13]: nit: unused import

File: contracts/PoolManager.sol [hunk 6]
```diff
@@ -108,6 +108,25 @@ contract PoolManager is IPoolManager, Fees, NoDelegateCall, ERC1155, IERC1155Rec
         return pools[id].positions.get(owner, tickLower, tickUpper);
     }
 
+    function _accountNoOp(PoolKey memory key, BalanceDelta hookDelta) internal {
+        // NOTE: we are in the lock state of the user, not the hook, unless the hook is the user
+        _accountDelta(key.currency0, hookDelta.amount0());
```
Comment (ewilz) [inline contracts/PoolManager.sol:113]: nit: why not just use `_accountPoolBalanceDelta` for the whole hookDelta?

File: contracts/PoolManager.sol [hunk 7]
```diff
@@ -108,6 +108,25 @@ contract PoolManager is IPoolManager, Fees, NoDelegateCall, ERC1155, IERC1155Rec
         return pools[id].positions.get(owner, tickLower, tickUpper);
     }
 
+    function _accountNoOp(PoolKey memory key, BalanceDelta hookDelta) internal {
+        // NOTE: we are in the lock state of the user, not the hook, unless the hook is the user
```
Comment (ewilz) [inline contracts/PoolManager.sol:112]: I think it'd be neater to take comments out of the function, they feel a bit inconsistent, and the code feels self-exlanatory, and just write one comment above along the lines of:
```
/// @dev on noOp, rebalance 1155 balances of hook and account deltas to the locker
```

File: contracts/test/GracefulReturnTestHooks.sol [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+    function setSwapFee(PoolKey calldata key, uint8 value) external {
+        swapFees[key.toId()] = value;
+    }
+
+    function setWithdrawFee(PoolKey calldata key, uint8 value) external {
+        withdrawFees[key.toId()] = value;
+    }
+
+    function simulateHead(address caller, uint256 nonce, uint256 stage, bytes32 head) external pure returns (bytes32) {
```
Comment (ewilz) [inline contracts/test/GracefulReturnTestHooks.sol:172]: i must say..... i am very confused

File: test/foundry-tests/PoolManager.t.sol [hunk 9]
```diff
@@ -984,6 +986,153 @@ contract PoolManagerTest is Test, Deployers, TokenFixture, GasSnapshot, IERC1155
...
+
+        // Donate
+        donateRouter.donate(key, 0, 0);
+
+        // should have donated 1 eth based on hook donate logic
+        assertEq(manager.balanceOf(hookAddr, CurrencyLibrary.toId(currency0)), 12 ether);
+        assertEq(manager.balanceOf(hookAddr, CurrencyLibrary.toId(currency1)), 9 ether);
+    }
+
+    function testPoolManagerGracefulReturnsOnHook(uint160 sqrtPriceX96) public {
```
Comment (ewilz) [inline test/foundry-tests/PoolManager.t.sol:1051]: I'd say this description does not explain what the test is testing, but just saying that it's testing a test contract. IMO tests should generally be very targeted (like the majority of the tests in this file) and only in certain special cases a test can be very comprehensive and go through many many things. I'm not completely following this whole GracefulReturns style testing.... it seems a bit confusing and unreadable to me

File: yarn.lock [hunk 10]
```diff
@@ -1625,10 +1625,10 @@
     "@types/sinon-chai" "^3.2.3"
     "@types/web3" "1.0.19"
 
-"@openzeppelin/contracts@4.4.2":
-  version "4.4.2"
-  resolved "https://registry.yarnpkg.com/@openzeppelin/contracts/-/contracts-4.4.2.tgz#4e889c9c66e736f7de189a53f8ba5b8d789425c2"
-  integrity sha512-NyJV7sJgoGYqbtNUWgzzOGW4T6rR19FmX1IJgXGdapGPWsuMelGJn9h03nos0iqfforCbCB0iYIR0MtIuIFLLw==
+"@openzeppelin/contracts@^4.9.2":
```
Comment (ewilz) [inline yarn.lock:1628]: por que?

File: package.json [hunk 11]
```diff
@@ -27,7 +27,7 @@
     "node": ">=10"
   },
   "dependencies": {
-    "@openzeppelin/contracts": "4.4.2"
+    "@openzeppelin/contracts": "^4.9.2"
```
Comment (ewilz) [inline package.json:30]: why?

File: contracts/PoolManager.sol [hunk 12]
```diff
@@ -108,6 +108,25 @@ contract PoolManager is IPoolManager, Fees, NoDelegateCall, ERC1155, IERC1155Rec
         return pools[id].positions.get(owner, tickLower, tickUpper);
     }
 
+    function _accountNoOp(PoolKey memory key, BalanceDelta hookDelta) internal {
```
Comment (ewilz) [inline contracts/PoolManager.sol:111]: nit: this function would be more well organized below `_accountPoolBalanceDelta` or at least in that general vicinity

File: contracts/test/GracefulReturnTestHooks.sol (hunk 8 above)
Comment (emmaguo13) [reply contracts/test/GracefulReturnTestHooks.sol:172]: This is all from #291, I can remove it after I merge my fee changes, the test is not super necessary

File: contracts/PoolManager.sol (hunk 6 above)
Comment (emmaguo13) [reply contracts/PoolManager.sol]: updated in most current commit, but it increases gas for modifyPosition by 56 and swap by 17

File: contracts/test/GracefulReturnTestHooks.sol (hunk 8 above)
Comment (thogard785) [reply contracts/test/GracefulReturnTestHooks.sol]: The original idea test concept that prompted my PR was using a sequencer inside the V4 hook, so some of that logic carried forward into the test.  My thinking was that sequencing the storage writes and using that to verify that *all* storage writes were successfully saved in the right order would help future-proof the test for transient storage... but i doubt it'll actually be relevant.

**Comment by snreynolds** [comment]
Closing in favor of this NoOp impl https://github.com/Uniswap/v4-core/pull/324

**Pull request closed without merging**